src/udx.c

```diff
@@ -874,11 +874,37 @@ trigger_send_callback (udx_socket_t *socket, udx_packet_t *pkt) {
 static void
 on_uv_poll (uv_poll_t *handle, int status, int events) {
   udx_socket_t *socket = handle->data;
+  ssize_t size;
 
-  if (socket->send_queue.len > 0 && events & UV_WRITABLE) {
-    udx_packet_t *pkt = (udx_packet_t *) udx__fifo_shift(&(socket->send_queue));
+  if (events & UV_READABLE) {
+    struct sockaddr_storage addr;
+    int addr_len = sizeof(addr);
+    uv_buf_t buf;
 
-    if (pkt == NULL) return;
+    memset(&addr, 0, addr_len);
+
+    char b[2048];
+    buf.base = (char *) &b;
+    buf.len = 2048;
+
+    while ((size = udx__recvmsg(socket, &buf, (struct sockaddr *) &addr, addr_len)) >= 0) {
+      if (!process_packet(socket, b, size, (struct sockaddr *) &addr) && socket->on_recv != NULL) {
+        buf.len = size;
+
+        if (is_addr_v4_mapped((struct sockaddr *) &addr)) {
+          addr_to_v4((struct sockaddr_in6 *) &addr);
+        }
+
+        socket->on_recv(socket, size, &buf, (struct sockaddr *) &addr);
+      }
+
+      buf.len = 2048;
+    }
+  }
+
+  while (socket->send_queue.len > 0 && events & UV_WRITABLE) {
```

Comment:
```suggestion
  while (events & UV_WRITABLE && socket->send_queue.len > 0) {
```

Feels a little more intuitive to me, not sure if you agree 😄

Reply:
Yuh agreed!